_posts/2020-04-27-ensuring-resolvers-aren't-rejected.md

```diff
@@ -0,0 +1,77 @@
+---
+layout: post
+title: Ensuring Resolvers Aren't Rejected
+date: 2020-04-27
+tags: appsync vtl resolver mapping template testing tdd
+author: Rick Foxcroft
+image: img/appsync.png
+---
+
+If you're using AppSync, you will no doubt have come across resolver mapping templates. To paraphrase the AWS [docs](https://docs.aws.amazon.com/appsync/latest/devguide/resolver-mapping-template-reference-overview.html), resolvers link a GraphQL query or mutation via a mapping template to a datasource and a mapping template is a way to indicate to AppSync how to translate an incoming GraphQL request into instructions that can be used to operate on your datasource.
+
+Mapping Templates are written in Apache Velocity Template Language ([VTL](https://velocity.apache.org/engine/1.7/user-guide.html)) and take the GraphQL request as input and produce a JSON document containing the resolver instructions as output. Mapping templates are great because they allow you to interact directly with AWS datastores such as DynamoDB, ElasticSearch and RDS. Allowing you to capture simple backend CRUD operations as configuration without the need for managing any related code.
+
+Except, they really can be code sometimes.
```

Comment:
Is this saying that it may not always be a VTL as there are other resolver types (i.e. lambda) or were you going to say that VTL really can be a hard to code?

Reply:
Even for a lambda resolver, one would still need vtl request & response mapping templates.

I think it's mentioned in a line following after this, but I was trying to get at how VTL is somewhat of a mini programming language, as it has loops, conditionals, variables etc.

Reply:
gotcha